File: rosjava_benchmarks/src/main/java/org/ros/rosjava_benchmarks/PubsubBenchmark.java
/*
 * Copyright (C) 2012 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package org.ros.rosjava_benchmarks;

import org.ros.concurrent.CancellableLoop;
import org.ros.concurrent.Rate;
import org.ros.concurrent.WallTimeRate;
import org.ros.message.Duration;
import org.ros.message.MessageListener;
import org.ros.message.Time;
import org.ros.namespace.GraphName;
import org.ros.node.AbstractNodeMain;
import org.ros.node.ConnectedNode;
import org.ros.node.topic.Publisher;
import org.ros.node.topic.Subscriber;
import tf.tfMessage;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author damonkohler@google.com (Damon Kohler)
 */
public class PubsubBenchmark extends AbstractNodeMain {

  private final AtomicInteger counter;

  private Publisher<std_msgs.String> statusPublisher;
  private Publisher<tf.tfMessage> tfPublisher;
  private Subscriber<tf.tfMessage> tfSubscriber;
  private Time time;

  public PubsubBenchmark() {
    counter = new AtomicInteger();
  }

  @Override
  public GraphName getDefaultNodeName() {
    return GraphName.of("pubsub_benchmark");
  }

  @Override
  public void onStart(final ConnectedNode connectedNode) {
    tfSubscriber = connectedNode.newSubscriber("tf", tf.tfMessage._TYPE);
    tfSubscriber.addMessageListener(new MessageListener<tf.tfMessage>() {
      @Override
      public void onNewMessage(tfMessage message) {
        counter.incrementAndGet();
      }
    });

    tfPublisher = connectedNode.newPublisher("tf", tf.tfMessage._TYPE);
    final tf.tfMessage tfMessage = tfPublisher.newMessage();
    geometry_msgs.TransformStamped transformStamped =
        connectedNode.getTopicMessageFactory().newFromType(geometry_msgs.TransformStamped._TYPE);
    tfMessage.getTransforms().add(transformStamped);
    connectedNode.executeCancellableLoop(new CancellableLoop() {
      @Override
      protected void loop() throws InterruptedException {
        tfPublisher.publish(tfMessage);
      }
    });

    time = connectedNode.getCurrentTime();
    statusPublisher = connectedNode.newPublisher("status", std_msgs.String._TYPE);
    final Rate rate = new WallTimeRate(1);
    final std_msgs.String status = statusPublisher.newMessage();
    connectedNode.executeCancellableLoop(new CancellableLoop() {
      @Override
      protected void loop() throws InterruptedException {
        Time now = connectedNode.getCurrentTime();
        Duration delta = now.subtract(time);
        if (delta.totalNsecs() > TimeUnit.NANOSECONDS.convert(5, TimeUnit.SECONDS)) {
          double hz = counter.getAndSet(0) * 1e9 / delta.totalNsecs();
          status.setData(String.format("%.2f Hz", hz));
          statusPublisher.publish(status);
          time = now;
        }
        rate.sleep();
      }
    });
  }
}

File: rosjava_benchmarks/scripts/pubsub_benchmark.py
#!/usr/bin/python
#
# Copyright (C) 2012 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

import threading

import roslib; roslib.load_manifest('rosjava_benchmarks')
import rospy

import tf.msg as tf_msgs
import std_msgs.msg as std_msgs


class PubsubBenchmark(object):

  def __init__(self):
    self.count = itertools.count()
    self.time = rospy.Time.now()
    self.lock = threading.Lock()

  def callback(self, _):
    with self.lock:
      next(self.count)

  def run(self):
    tf_publisher = rospy.Publisher('tf', tf_msgs.tfMessage)
    status_publisher = rospy.Publisher('status', std_msgs.String)
    rospy.Subscriber('tf', tf_msgs.tfMessage, self.callback)
    while not rospy.is_shutdown():
      tf_publisher.publish(tf_msgs.tfMessage())
      if (rospy.Time.now() - self.time) > rospy.Duration(5):
        status_publisher.publish(std_msgs.String('%.2f Hz' % (next(self.count) / 5.0)))
        with self.lock:
          self.count = itertools.count()
          self.time = rospy.Time.now()


if __name__ == '__main__':
  try:
    rospy.init_node('benchmark')
    PubsubBenchmark().run()
  except rospy.ROSInterruptException:
    pass
